File: src/java/logic/Board.java
package logic;

import java.util.ArrayList;
import java.util.Arrays;

public class Board {

    //round 43 is max

    //two rules for creating a board: width must not exceed 9, winCondition must not exceed width AND height

    private String[][] board; //height, width
    private String currentTurn;
    private int winCondition;
    private int turn;
    public int lastX;
    public int lastY;
    private static final int MAX_WIDTH = 9;
    public boolean playing = true;

    public Board(int width, int height, int winCondition) {
        this.board = new String[height][width];
        for (int i = 0; i < height; i++) {
            for (int u = 0; u < width; u++) {
                board[i][u] = "_";
            }
        }
    
        this.turn = 1;
        this.currentTurn = "X";
        this.winCondition = winCondition;
    }

    public Board(Board newBoard) { //alternate constructor for the ai player
        int width = newBoard.getWidth();
        int height = newBoard.getHeight();
        this.board = new String[height][width];

        for (int i = 0; i < height; i++) {
            this.board[i] = Arrays.copyOf(newBoard.board[i], width);
        }

        this.currentTurn = newBoard.currentTurn;
        this.winCondition = newBoard.winCondition;
        this.turn = newBoard.turn;
        this.lastX = newBoard.lastX;
        this.lastY = newBoard.lastY;
    }

    public void drop(int col, String piece, boolean isSimulated) {
        if (col < getWidth() && col >= 0) {
            for (int i = getHeight() - 1; i >= 0; i--) {
                if (board[i][col].equals("_")) {
                    board[i][col] = piece;
                    lastX = col;
                    lastY = i;

                    if (!checkWin(lastX, lastY).equals("no winner") && !isSimulated) {
                        endGame();
                    }

                    if (!isSimulated) {
                        changeTurn();
                        turn++;
                    }
                    break;
                }
            }
        }
    }

    public void endGame() {
        if (currentTurn == "X") console.log("You win!");
        if (currentTurn == "O") console.log("You lose!");
        playing = false;
    }

    public String checkWin(int x, int y) {
        //horizontal
        for (int i = (Math.max(0, x-winCondition+1)); i <= (Math.min(getWidth()-winCondition, x)); i++) {
            String currentVal = board[y][i];
            boolean matches = true;
            for (int u = 1; u < winCondition; u++) {
                String checkedVal = board[y][i+u];
                if (!currentVal.equals(checkedVal)) {
                    matches = false;
                    break;
                }
            }
            if (matches == true && !currentVal.equals("_")) {
                return currentVal;
            }
        }

        //vertical
        for (int i = Math.max(0, y-winCondition+1); i <= Math.min(getHeight()-winCondition, y); i++) {
            String currentVal = board[i][x];
            boolean matches = true;
            for (int u = 1; u < winCondition; u++) {
                String checkedVal = board[i+u][x];
                if (!currentVal.equals(checkedVal)) {
                    matches = false;
                    break;
                }
            }
            if (matches == true && !currentVal.equals("_")) {
                return currentVal;
            }
        }

        //diagonal, up to down left to right
        for (int i = -winCondition + 1; i <= 0; i++) {
            //start at wincondition spaces diagonal to space we're checking
            int startX = x + i; 
            int startY = y + i;
            //check outer bounds (top left, bottom right bounds)
            if (startX < 0 || 
                startY < 0 || 
                startX + winCondition > getWidth() || 
                startY + winCondition > getHeight()) {
                continue; //skip checking it, as it's out of bounds
            }
            
            String currentVal = board[startY][startX];
            boolean matches = true;
            for (int u = 1; u < winCondition; u++) {
                String checkedVal = board[startY + u][startX + u];
                if (!currentVal.equals(checkedVal)) {
                    matches = false;
                    break;
                }
            }
            if (matches == true && !currentVal.equals("_")) {
                return currentVal;
            }
        }

        //anti-diagonal, down to up left to right
        for (int i = -winCondition + 1; i <= 0; i++) {
            //start at wincondition spaces diagonal to space we're checking
            int startX = x + i;
            int startY = y - i;
            //check if within outer bounds
            if (startX < 0 || 
                startY >= getHeight() || 
                startX + winCondition > getWidth() || 
                startY - winCondition + 1 < 0) {
                continue; //out of bounds
            }

            String currentVal = board[startY][startX];
            boolean matches = true;
            for (int u = 1; u < winCondition; u++) {
                String checkedVal = board[startY - u][startX + u];
                if (!currentVal.equals(checkedVal)) {
                    matches = false;
                    break;
                }
            }
            if (matches && !currentVal.equals("_")) {
                return currentVal;
            }
        }

        return "no winner";
    }

    public void changeTurn() {
        currentTurn = (currentTurn.equals("X")) ? "O" : "X";
    }

    public int getWidth() {
        return board[0].length;
    }

    public int getHeight() {
        return board.length;
    }

    public String[][] getBoard() {
        return board;
    }
    
    public String getCurrentTurn() {
        return currentTurn;
    }

    public int getWinCondition() {
        return winCondition;
    }

    public int getTurn() {
        return turn;
    }
    
    public boolean isFull() {
        return turn > getHeight() * getWidth();
    }
    
    public boolean isColumnOpen(int col) {
        return board[0][col].equals("_");
    }

    public String toString() {
        console.cls();
        String result = "VS. " + ConnectFour.opponent + "\n\n";
        for (String[] i : board) {
            result += "| ";
            for (String u : i) {
                result += u + " | ";
            }
            result += "\n";
        }
        result += "\n";
        for (int i = 1; i < getWidth() + 1; i++) {
            result += "  " + i + " ";
        }
        console.log("Turn: " + turn + ", Winner: " + checkWin(lastX, lastY));
        return result;
    }
}

File: src/main/frontend/src/board.ts
export class Board {
    width: number;
    height: number;
    winCondition: number;
    board: string[][];

    constructor(width: number = 7, height: number = 6, winCondition: number = 4) {
        this.width = width;
        this.height = height;
        this.winCondition = winCondition;
        this.board = Array.from({ length: height }, () =>
            new Array(width).fill('_')
        );
    }

    clone(): Board {
        const newBoard = new Board(this.width, this.height, this.winCondition);
        for (let r = 0; r < this.height; r++) newBoard.board[r] = this.board[r].slice();
        return newBoard;
    }

    isFull(): boolean {
        for (let col = 0; col < this.width; col++) {
            if (this.board[0][col] === '_') {
                return false;
            }
        }
        return true;
    }

    getValidMoves(): number[] {
        const res: number[] = [];
        for (let col = 0; col < this.width; col++) {
            if (this.board[0][col] === '_') {
                res.push(col);
            }
        }
        return res;
    }

    drop(col: number, piece: string): void {
        for (let row = this.height - 1; row >= 0; row--) {
            if (this.board[row][col] === '_') {
                this.board[row][col] = piece;
                break;
            }
        }
    }

    getAt(row: number, col: number): string {
        return this.board[row][col];
    }

    // Returns 'X' or 'O' or 'no winner'
    checkWin(): string {
        // Local aliases for readability
        const board = this.board;
        const rows = this.height;
        const cols = this.width;
        const winCon = this.winCondition;

        // check if piece is in bounds
        const inBounds = (row: number, col: number) => {
            return (
                row >= 0 && //above bottom
                row < rows && //below top
                col >= 0 && //left of right
                col < cols //right of left
            );
        };

        // check for streak in a direction starting at (row, col)
        function checkDirection(row: number, col: number, deltaRow: number, deltaCol: number): boolean {
            const startPiece = board[row][col];
            
            if (startPiece === '_') return false; // empty cell can't form a win
            for (let step = 1; step < winCon; step++) {
                const newRow = row + deltaRow * step;
                const newCol = col + deltaCol * step;
                if (
                    !inBounds(newRow, newCol) || //check if it's out of bounds
                    board[newRow][newCol] !== startPiece //or if it's not the same as the starting piece
                ) {
                    return false;
                }
            }
            return true;
        }

        // check for win
        for (let row = 0; row < rows; row++) {
            for (let col = 0; col < cols; col++) {
                if (board[row][col] === '_') continue;
                if (
                    checkDirection(row, col, 0, 1) ||   // horizontal
                    checkDirection(row, col, 1, 0) ||   // vertical
                    checkDirection(row, col, 1, 1) ||   // diagonal down
                    checkDirection(row, col, -1, 1)     // diagonal up
                ) {
                    return board[row][col];
                }
            }
        }

        // no winner
        return 'no winner';
    }
}

